File: backend/api/perfiles_engine.php
<?php
/**
 * Wrapper de API para el Motor transaccional en Python de Perfiles y Permisos (NOFTRAB)
 */

header('Content-Type: application/json; charset=utf-8');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: POST, GET, PUT, DELETE, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Authorization');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    http_response_code(200);
    exit;
}

require_once '../config.php';

session_start();

// Validar token de autorización si no hay sesión PHP activa
$bearer_token = null;
$auth_header = $_SERVER['HTTP_AUTHORIZATION'] ?? '';
if (empty($auth_header) && function_exists('apache_request_headers')) {
    $headers = apache_request_headers();
    $auth_header = $headers['Authorization'] ?? $headers['authorization'] ?? '';
}
if (preg_match('/Bearer\s+(.+)$/i', $auth_header, $matches)) {
    $bearer_token = $matches[1];
}

if (!isset($_SESSION['usuario_id']) && empty($bearer_token)) {
    respuestaJSON(false, 'Sesión no válida', null, 401);
}

$usuario_actual = $_SESSION['usuario_id'] ?? 1;

// Solo el Administrador (perfil_id = 1) o alguien con permisos especiales puede gestionar esto
$perfil_usuario = obtenerPerfilUsuario($usuario_actual);
if ($usuario_actual != 1 && $perfil_usuario['id'] != 1 && !tienePermiso($usuario_actual, 'CONF_TOTAL') && !tienePermiso($usuario_actual, 'PER_GESTIONAR')) {
    respuestaJSON(false, 'Acceso denegado: Se requiere perfil de Administrador.', null, 403);
}

$metodo = $_SERVER['REQUEST_METHOD'];
$request_uri = $_SERVER['REQUEST_URI'];
$script_name = $_SERVER['SCRIPT_NAME'];

// Obtener la ruta relativa al script
$ruta = $_SERVER['PATH_INFO'] ?? '';
if (empty($ruta)) {
    $pos = strpos($request_uri, $script_name);
    if ($pos !== false) {
        $ruta = substr($request_uri, $pos + strlen($script_name));
    }
}

$partes = explode('/', trim($ruta, '/'));
$endpoint = $partes[0] ?? '';

$python_script = dirname(__DIR__) . '/perfiles_engine.py';

// Función para ejecutar comando python
function runPythonCommand($cmd, $args = []) {
    $python_bin = 'python'; // Por defecto
    
    // Si conocemos la ruta de Python de WAMP podemos usarla o simplemente 'python'
    $full_cmd = $python_bin . ' ' . escapeshellarg($cmd);
    foreach ($args as $arg) {
        $full_cmd .= ' ' . escapeshellarg($arg);
    }
    
    $output = [];
    $return_var = 0;
    exec($full_cmd . ' 2>&1', $output, $return_var);
    
    $output_str = implode("\n", $output);
    $json = json_decode($output_str, true);
    
    if (json_last_error() !== JSON_ERROR_NONE) {
        return [
            'exito' => false,
            'mensaje' => 'Error al invocar el motor en Python',
            'error_detalle' => $output_str
        ];
    }
    
    return $json;
}

try {
    if ($endpoint === 'listar' && $metodo === 'GET') {
        // Listar módulos y funciones
        $res = runPythonCommand($python_script, ['list_modules_functions']);
        respuestaJSON($res['exito'], $res['mensaje'], $res['datos'] ?? null);
        
    } elseif ($endpoint === 'obtener' && $metodo === 'GET') {
        // Obtener permisos de un perfil específico
        $perfil_id = intval($partes[1] ?? 0);
        if ($perfil_id === 0) {
            respuestaJSON(false, 'ID de perfil no proporcionado', null, 400);
        }
        
        $res = runPythonCommand($python_script, ['get_profile_permissions', $perfil_id]);
        respuestaJSON($res['exito'], $res['mensaje'], $res['datos'] ?? null);
        
    } elseif ($endpoint === 'guardar' && $metodo === 'POST') {
        // Guardar permisos granulares para un perfil (PHP nativo para evitar cuelgues de Apache en Windows)
        $perfil_id = intval($partes[1] ?? 0);
        if ($perfil_id === 0) {
            respuestaJSON(false, 'ID de perfil no proporcionado', null, 400);
        }
        
        $json_input = file_get_contents('php://input');
        $permisos = json_decode($json_input, true);
        
        if (!is_array($permisos)) {
            respuestaJSON(false, 'Formato de permisos inválido', null, 400);
        }
        
        $db = Database::getInstance()->getConnection();
        
        // 1. Obtener el perfil
        $stmt_perfil = $db->prepare("SELECT nombre_perfil FROM perfiles WHERE id = ?");
        $stmt_perfil->bind_param("i", $perfil_id);
        $stmt_perfil->execute();
        $res_perfil = $stmt_perfil->get_result()->fetch_assoc();
        $stmt_perfil->close();
        
        if (!$res_perfil) {
            respuestaJSON(false, 'Perfil no encontrado', null, 404);
        }
        
        // 2. Obtener permisos antiguos para la auditoría (NOFTRAB)
        $permisos_antiguos = [];
        $stmt_ant = $db->prepare("SELECT * FROM permisos_perfil WHERE perfil_id = ?");
        $stmt_ant->bind_param("i", $perfil_id);
        $stmt_ant->execute();
        $res_ant = $stmt_ant->get_result();
        while ($row = $res_ant->fetch_assoc()) {
            $permisos_antiguos[] = $row;
        }
        $stmt_ant->close();
        
        // Iniciar transacción
        $db->begin_transaction();
        
        try {
            // 3. Eliminar permisos antiguos
            $stmt_del = $db->prepare("DELETE FROM permisos_perfil WHERE perfil_id = ?");
            $stmt_del->bind_param("i", $perfil_id);
            $stmt_del->execute();
            $stmt_del->close();
            
            // 4. Insertar nuevos permisos granulares
            $sql_insert = "INSERT INTO permisos_perfil (
                perfil_id, funcion_id, modulo_id, puede_ejecutar, ver_datos,
                crear_datos, editar_datos, eliminar_datos, ver_reportes,
                exportar_datos, importar_datos, imprimir_datos, solo_propios, creado_por
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
            
            $stmt_ins = $db->prepare($sql_insert);
            if (!$stmt_ins) {
                throw new Exception("Error al preparar la inserción de permisos.");
            }
            
            foreach ($permisos as $p) {
                $f_id = intval($p['funcion_id']);
                $m_id = intval($p['modulo_id']);
                $p_ejec = isset($p['puede_ejecutar']) && $p['puede_ejecutar'] ? 1 : 0;
                $v_dat = isset($p['ver_datos']) && $p['ver_datos'] ? 1 : 0;
                $c_dat = isset($p['crear_datos']) && $p['crear_datos'] ? 1 : 0;
                $e_dat = isset($p['editar_datos']) && $p['editar_datos'] ? 1 : 0;
                $el_dat = isset($p['eliminar_datos']) && $p['eliminar_datos'] ? 1 : 0;
                $v_rep = isset($p['ver_reportes']) && $p['ver_reportes'] ? 1 : 0;
                $ex_dat = isset($p['exportar_datos']) && $p['exportar_datos'] ? 1 : 0;
                $im_dat = isset($p['importar_datos']) && $p['importar_datos'] ? 1 : 0;
                $imp_dat = isset($p['imprimir_datos']) && $p['imprimir_datos'] ? 1 : 0;
                $s_prop = isset($p['solo_propios']) && $p['solo_propios'] ? 1 : 0;
                
                $stmt_ins->bind_param(
                    "iiiiiiiiiiiiii",
                    $perfil_id, $f_id, $m_id, $p_ejec, $v_dat,
                    $c_dat, $e_dat, $el_dat, $v_rep,
                    $ex_dat, $im_dat, $imp_dat, $s_prop, $usuario_actual
                );
                $stmt_ins->execute();
            }
            $stmt_ins->close();
            
            // 5. Escribir auditoría de accesos (NOFTRAB)
            $sql_audit = "INSERT INTO auditoria_accesos (
                usuario_id, tipo_evento, modulo_accedido, funcion_ejecutada,
                descripcion_evento, direccion_ip, navegador_user_agent,
                resultado, tabla_afectada, registro_afectado_id,
                operacion_realizada, valor_anterior, valor_nuevo
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
            
            $stmt_aud = $db->prepare($sql_audit);
            if ($stmt_aud) {
                $tipo_ev = 'cambio_permiso';
                $mod_acc = 'configuracion';
                $func_ej = 'PER_ASIGNAR';
                $descr = "Actualización granular de permisos para el perfil: " . $res_perfil['nombre_perfil'];
                $ip_addr = $_SERVER['REMOTE_ADDR'] ?? '127.0.0.1';
                $u_agent = $_SERVER['HTTP_USER_AGENT'] ?? 'PHP Engine API';
                $res_ok = 'exitoso';
                $t_afect = 'permisos_perfil';
                $op_real = 'update';
                $val_ant = json_encode($permisos_antiguos);
                $val_nuev = json_encode($permisos);
                
                $stmt_aud->bind_param(
                    "issssssssisss",
                    $usuario_actual, $tipo_ev, $mod_acc, $func_ej,
                    $descr, $ip_addr, $u_agent, $res_ok, $t_afect,
                    $perfil_id, $op_real, $val_ant, $val_nuev
                );
                $stmt_aud->execute();
                $stmt_aud->close();
            }
            
            // Confirmar transacción
            $db->commit();
            respuestaJSON(true, 'Permisos guardados y auditados exitosamente', null);
            
        } catch (Exception $e) {
            $db->rollback();
            respuestaJSON(false, 'Error al guardar permisos: ' . $e->getMessage(), null, 500);
        }
        
    } elseif ($endpoint === 'crear' && $metodo === 'POST') {
        // Crear un nuevo perfil
        $json_input = file_get_contents('php://input');
        
        $temp_file = dirname(__DIR__) . '/logs/temp_crear_perfil_' . time() . '.json';
        file_put_contents($temp_file, $json_input);
        
        $res = runPythonCommand($python_script, ['create_profile', $temp_file, $usuario_actual]);
        
        if (file_exists($temp_file)) {
            unlink($temp_file);
        }
        
        respuestaJSON($res['exito'], $res['mensaje'], $res['perfil_id'] ?? null, $res['exito'] ? 201 : 400);
        
    } elseif ($endpoint === 'editar' && ($metodo === 'PUT' || $metodo === 'POST')) {
        // Editar un perfil existente
        $perfil_id = intval($partes[1] ?? 0);
        if ($perfil_id === 0) {
            respuestaJSON(false, 'ID de perfil no proporcionado', null, 400);
        }
        
        $json_input = file_get_contents('php://input');
        
        $temp_file = dirname(__DIR__) . '/logs/temp_editar_perfil_' . $perfil_id . '_' . time() . '.json';
        file_put_contents($temp_file, $json_input);
        
        $res = runPythonCommand($python_script, ['update_profile', $perfil_id, $temp_file, $usuario_actual]);
        
        if (file_exists($temp_file)) {
            unlink($temp_file);
        }
        
        respuestaJSON($res['exito'], $res['mensaje'], null);
        
    } elseif ($endpoint === 'eliminar' && ($metodo === 'DELETE' || $metodo === 'POST')) {
        // Eliminar perfil
        $perfil_id = intval($partes[1] ?? 0);
        if ($perfil_id === 0 && isset($_GET['id'])) {
            $perfil_id = intval($_GET['id']);
        }
        
        if ($perfil_id === 0) {
            respuestaJSON(false, 'ID de perfil no proporcionado', null, 400);
        }
        
        $res = runPythonCommand($python_script, ['delete_profile', $perfil_id, $usuario_actual]);
        respuestaJSON($res['exito'], $res['mensaje'], null);
        
    } elseif ($endpoint === 'copiar' && $metodo === 'POST') {
        // Copiar permisos de un perfil a otro
        $json_input = file_get_contents('php://input');
        $data = json_decode($json_input, true);
        $origen_id = intval($data['origen_id'] ?? 0);
        $destino_id = intval($data['destino_id'] ?? 0);
        
        if ($origen_id === 0 || $destino_id === 0) {
            respuestaJSON(false, 'Se requiere perfil de origen y destino', null, 400);
        }
        
        $res = runPythonCommand($python_script, ['copy_profile_permissions', $origen_id, $destino_id, $usuario_actual]);
        respuestaJSON($res['exito'], $res['mensaje'], null);
        
    } else {
        respuestaJSON(false, "Endpoint '$endpoint' no encontrado", null, 404);
    }
} catch (Exception $e) {
    respuestaJSON(false, 'Error en Wrapper API: ' . $e->getMessage(), null, 500);
}
?>

File: backend/perfiles_engine.py
import sys
sys.path.append(r"C:\Users\jhenr\AppData\Roaming\Python\Python314\site-packages")
sys.path.append(r"C:\Users\jhenr\AppData\Local\Programs\Python\Python314\Lib\site-packages")
import json
import pymysql
import pymysql.cursors

import os

# Configuración de base de datos dinámica para Windows (WAMP) y VPS Linux
DB_USER = os.getenv('DB_USER', 'masque_user')
DB_PASS = os.getenv('DB_PASS', 'MasQF_2026_Secure!')
cwd = os.getcwd().replace('\\', '/')
DB_NAME = os.getenv('DB_NAME', 'masque_fianzas_dev' if 'dev_plataforma' in cwd else 'masque_fianzas_integrada_01')

# Si estamos en Windows con WAMP sin masque_user, fallback a root
DB_CONFIG = {
    'host': '127.0.0.1',
    'user': DB_USER,
    'password': DB_PASS,
    'database': DB_NAME,
    'charset': 'utf8mb4',
    'cursorclass': pymysql.cursors.DictCursor
}

def get_connection():
    try:
        return pymysql.connect(**DB_CONFIG)
    except Exception as e:
        # Fallback a root sin password para WAMP local si masque_user no existe localmente
        alt_config = DB_CONFIG.copy()
        alt_config['user'] = 'root'
        alt_config['password'] = ''
        return pymysql.connect(**alt_config)

def load_json_param(param):
    import os
    if param.endswith('.json') and os.path.exists(param):
        try:
            with open(param, 'r', encoding='utf-8') as f:
                return f.read()
        except Exception:
            pass
    return param


def list_modules_functions():
    connection = get_connection()
    try:
        with connection.cursor() as cursor:
            # Obtener todos los módulos
            cursor.execute("SELECT id, nombre_modulo, nombre_modulo AS codigo_modulo, descripcion, estado FROM modulos ORDER BY id")
            modulos = cursor.fetchall()
            
            # Obtener todas las funciones
            cursor.execute("SELECT id, modulo_id, nombre_funcion, codigo_funcion, descripcion, tipo_permiso, estado FROM funciones_modulo ORDER BY modulo_id, id")
            funciones = cursor.fetchall()
            
            # Agrupar funciones por módulo
            modulos_dict = {m['id']: {**m, 'funciones': []} for m in modulos}
            for f in funciones:
                m_id = f['modulo_id']
                if m_id in modulos_dict:
                    modulos_dict[m_id]['funciones'].append(f)
                    
            return {
                'exito': True,
                'mensaje': 'Módulos y funciones obtenidos',
                'datos': list(modulos_dict.values())
            }
    except Exception as e:
        return {'exito': False, 'mensaje': f'Error en list_modules_functions: {str(e)}'}
    finally:
        connection.close()

def get_profile_permissions(perfil_id):
    connection = get_connection()
    try:
        perfil_id = int(perfil_id)
        with connection.cursor() as cursor:
            # Obtener información del perfil
            cursor.execute("SELECT * FROM perfiles WHERE id = %s", (perfil_id,))
            perfil = cursor.fetchone()
            if not perfil:
                return {'exito': False, 'mensaje': 'Perfil no encontrado'}
                
            # Obtener los permisos del perfil
            cursor.execute("SELECT * FROM permisos_perfil WHERE perfil_id = %s", (perfil_id,))
            permisos = cursor.fetchall()
            
            return {
                'exito': True,
                'mensaje': 'Permisos del perfil obtenidos',
                'perfil': perfil,
                'datos': permisos
            }
    except Exception as e:
        return {'exito': False, 'mensaje': f'Error en get_profile_permissions: {str(e)}'}
    finally:
        connection.close()

def save_profile_permissions(perfil_id, permisos_json, usuario_id):
    connection = get_connection()
    try:
        perfil_id = int(perfil_id)
        usuario_id = int(usuario_id)
        permisos = json.loads(load_json_param(permisos_json))
        
        with connection.cursor() as cursor:
            # Obtener el perfil
            cursor.execute("SELECT nombre_perfil FROM perfiles WHERE id = %s", (perfil_id,))
            perfil = cursor.fetchone()
            if not perfil:
                return {'exito': False, 'mensaje': 'Perfil no encontrado'}
            
            # Obtener permisos antiguos para la auditoría (NOFTRAB)
            cursor.execute("SELECT * FROM permisos_perfil WHERE perfil_id = %s", (perfil_id,))
            permisos_antiguos = cursor.fetchall()
            
            # Iniciar transacción
            connection.begin()
            
            # Eliminar permisos antiguos
            cursor.execute("DELETE FROM permisos_perfil WHERE perfil_id = %s", (perfil_id,))
            
            # Insertar nuevos permisos granulares
            sql_insert = """
                INSERT INTO permisos_perfil (
                    perfil_id, funcion_id, modulo_id, puede_ejecutar, ver_datos,
                    crear_datos, editar_datos, eliminar_datos, ver_reportes,
                    exportar_datos, importar_datos, imprimir_datos, solo_propios, creado_por
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            
            for p in permisos:
                cursor.execute(sql_insert, (
                    perfil_id,
                    int(p['funcion_id']),
                    int(p['modulo_id']),
                    1 if p.get('puede_ejecutar', True) else 0,
                    1 if p.get('ver_datos', False) else 0,
                    1 if p.get('crear_datos', False) else 0,
                    1 if p.get('editar_datos', False) else 0,
                    1 if p.get('eliminar_datos', False) else 0,
                    1 if p.get('ver_reportes', False) else 0,
                    1 if p.get('exportar_datos', False) else 0,
                    1 if p.get('importar_datos', False) else 0,
                    1 if p.get('imprimir_datos', False) else 0,
                    1 if p.get('solo_propios', False) else 0,
                    usuario_id
                ))
            
            # Escribir auditoría detallada de accesos (NOFTRAB)
            sql_audit = """
                INSERT INTO auditoria_accesos (
                    usuario_id, tipo_evento, modulo_accedido, funcion_ejecutada,
                    descripcion_evento, direccion_ip, navegador_user_agent,
                    resultado, tabla_afectada, registro_afectado_id,
                    operacion_realizada, valor_anterior, valor_nuevo
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            
            descripcion = f"Actualización granular de permisos para el perfil: {perfil['nombre_perfil']}"
            cursor.execute(sql_audit, (
                usuario_id,
                'cambio_permiso',
                'configuracion',
                'PER_ASIGNAR',
                descripcion,
                '127.0.0.1',
                'Python Engine API',
                'exitoso',
                'permisos_perfil',
                perfil_id,
                'update',
                json.dumps(permisos_antiguos, default=str),
                json.dumps(permisos, default=str)
            ))
            
            # Confirmar transacción
            connection.commit()
            
            return {
                'exito': True,
                'mensaje': 'Permisos guardados y auditados exitosamente'
            }
    except Exception as e:
        connection.rollback()
        return {'exito': False, 'mensaje': f'Error en save_profile_permissions: {str(e)}'}
    finally:
        connection.close()

def create_profile(profile_json, usuario_creador):
    connection = get_connection()
    try:
        usuario_creador = int(usuario_creador)
        datos = json.loads(load_json_param(profile_json))
        
        nombre_perfil = datos.get('nombre_perfil')
        descripcion = datos.get('descripcion', '')
        nivel_jerarquico = int(datos.get('nivel_jerarquico', 5))
        hereda_de = datos.get('hereda_de')
        
        if not nombre_perfil:
            return {'exito': False, 'mensaje': 'El nombre del perfil es requerido'}
            
        with connection.cursor() as cursor:
            # Validar nombre único
            cursor.execute("SELECT id FROM perfiles WHERE nombre_perfil = %s", (nombre_perfil,))
            if cursor.fetchone():
                return {'exito': False, 'mensaje': 'El nombre del perfil ya existe'}
                
            connection.begin()
            
            # Insertar perfil
            sql_insert = """
                INSERT INTO perfiles (
                    nombre_perfil, descripcion, nivel_jerarquico, estado,
                    es_predeterminado, hereda_de, creado_por
                ) VALUES (%s, %s, %s, 'activo', 0, %s, %s)
            """
            cursor.execute(sql_insert, (nombre_perfil, descripcion, nivel_jerarquico, hereda_de, usuario_creador))
            perfil_id = cursor.lastrowid
            
            # Si hereda de otro perfil, copiar sus permisos
            if hereda_de:
                cursor.execute("""
                    INSERT INTO permisos_perfil (
                        perfil_id, funcion_id, modulo_id, puede_ejecutar, ver_datos,
                        crear_datos, editar_datos, eliminar_datos, ver_reportes,
                        exportar_datos, importar_datos, imprimir_datos, solo_propios, creado_por
                    )
                    SELECT %s, funcion_id, modulo_id, puede_ejecutar, ver_datos,
                           crear_datos, editar_datos, eliminar_datos, ver_reportes,
                           exportar_datos, importar_datos, imprimir_datos, solo_propios, %s
                    FROM permisos_perfil WHERE perfil_id = %s
                """, (perfil_id, usuario_creador, int(hereda_de)))
            
            # Escribir auditoría
            sql_audit = """
                INSERT INTO auditoria_accesos (
                    usuario_id, tipo_evento, modulo_accedido, funcion_ejecutada,
                    descripcion_evento, direccion_ip, navegador_user_agent,
                    resultado, tabla_afectada, registro_afectado_id,
                    operacion_realizada
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            cursor.execute(sql_audit, (
                usuario_creador,
                'crear_perfil',
                'configuracion',
                'PER_GESTIONAR',
                f"Creado perfil {nombre_perfil} con ID {perfil_id}",
                '127.0.0.1',
                'Python Engine API',
                'exitoso',
                'perfiles',
                perfil_id,
                'insert'
            ))
            
            connection.commit()
            return {
                'exito': True,
                'mensaje': 'Perfil creado exitosamente',
                'perfil_id': perfil_id
            }
    except Exception as e:
        connection.rollback()
        return {'exito': False, 'mensaje': f'Error en create_profile: {str(e)}'}
    finally:
        connection.close()

def update_profile(perfil_id, profile_json, usuario_modificador):
    connection = get_connection()
    try:
        perfil_id = int(perfil_id)
        usuario_modificador = int(usuario_modificador)
        datos = json.loads(load_json_param(profile_json))
        
        nombre_perfil = datos.get('nombre_perfil')
        descripcion = datos.get('descripcion', '')
        nivel_jerarquico = int(datos.get('nivel_jerarquico', 5))
        estado = datos.get('estado', 'activo')
        
        with connection.cursor() as cursor:
            # Obtener estado anterior
            cursor.execute("SELECT * FROM perfiles WHERE id = %s", (perfil_id,))
            perfil_antiguo = cursor.fetchone()
            if not perfil_antiguo:
                return {'exito': False, 'mensaje': 'Perfil no encontrado'}
                
            connection.begin()
            
            # Actualizar perfil
            sql_update = """
                UPDATE perfiles SET
                    nombre_perfil = %s,
                    descripcion = %s,
                    nivel_jerarquico = %s,
                    estado = %s,
                    modificado_por = %s,
                    fecha_modificacion = NOW()
                WHERE id = %s
            """
            cursor.execute(sql_update, (nombre_perfil, descripcion, nivel_jerarquico, estado, usuario_modificador, perfil_id))
            
            # Registrar auditoría
            sql_audit = """
                INSERT INTO auditoria_accesos (
                    usuario_id, tipo_evento, modulo_accedido, funcion_ejecutada,
                    descripcion_evento, direccion_ip, navegador_user_agent,
                    resultado, tabla_afectada, registro_afectado_id,
                    operacion_realizada, valor_anterior, valor_nuevo
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            cursor.execute(sql_audit, (
                usuario_modificador,
                'editar_perfil',
                'configuracion',
                'PER_GESTIONAR',
                f"Modificado perfil {nombre_perfil} con ID {perfil_id}",
                '127.0.0.1',
                'Python Engine API',
                'exitoso',
                'perfiles',
                perfil_id,
                'update',
                json.dumps(perfil_antiguo, default=str),
                json.dumps(datos, default=str)
            ))
            
            connection.commit()
            return {
                'exito': True,
                'mensaje': 'Perfil actualizado exitosamente'
            }
    except Exception as e:
        connection.rollback()
        return {'exito': False, 'mensaje': f'Error en update_profile: {str(e)}'}
    finally:
        connection.close()

def delete_profile(perfil_id, usuario_borrador):
    connection = get_connection()
    try:
        perfil_id = int(perfil_id)
        usuario_borrador = int(usuario_borrador)
        
        with connection.cursor() as cursor:
            # Obtener datos del perfil antes de borrar
            cursor.execute("SELECT * FROM perfiles WHERE id = %s", (perfil_id,))
            perfil = cursor.fetchone()
            if not perfil:
                return {'exito': False, 'mensaje': 'Perfil no encontrado'}
                
            # No permitir borrar el Administrador (ID 1)
            if perfil_id == 1:
                return {'exito': False, 'mensaje': 'No se puede eliminar el perfil Administrador'}
                
            connection.begin()
            
            # Eliminar permisos asociados
            cursor.execute("DELETE FROM permisos_perfil WHERE perfil_id = %s", (perfil_id,))
            
            # Eliminar perfil
            cursor.execute("DELETE FROM perfiles WHERE id = %s", (perfil_id,))
            
            # Registrar auditoría
            sql_audit = """
                INSERT INTO auditoria_accesos (
                    usuario_id, tipo_evento, modulo_accedido, funcion_ejecutada,
                    descripcion_evento, direccion_ip, navegador_user_agent,
                    resultado, tabla_afectada, registro_afectado_id,
                    operacion_realizada, valor_anterior
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            cursor.execute(sql_audit, (
                usuario_borrador,
                'eliminar_perfil',
                'configuracion',
                'PER_GESTIONAR',
                f"Eliminado perfil {perfil['nombre_perfil']} con ID {perfil_id}",
                '127.0.0.1',
                'Python Engine API',
                'exitoso',
                'perfiles',
                perfil_id,
                'delete',
                json.dumps(perfil, default=str)
            ))
            
            connection.commit()
            return {
                'exito': True,
                'mensaje': 'Perfil eliminado exitosamente'
            }
    except Exception as e:
        connection.rollback()
        return {'exito': False, 'mensaje': f'Error en delete_profile: {str(e)}'}
    finally:
        connection.close()

def copy_profile_permissions(origen_id, destino_id, usuario_id):
    connection = get_connection()
    try:
        origen_id = int(origen_id)
        destino_id = int(destino_id)
        usuario_id = int(usuario_id)
        
        with connection.cursor() as cursor:
            # Validar perfiles
            cursor.execute("SELECT nombre_perfil FROM perfiles WHERE id = %s", (origen_id,))
            perfil_origen = cursor.fetchone()
            cursor.execute("SELECT nombre_perfil FROM perfiles WHERE id = %s", (destino_id,))
            perfil_destino = cursor.fetchone()
            
            if not perfil_origen or not perfil_destino:
                return {'exito': False, 'mensaje': 'Perfil de origen o destino no encontrado'}
                
            if destino_id == 1:
                return {'exito': False, 'mensaje': 'No se pueden sobrescribir los permisos del Administrador principal'}
                
            # Obtener permisos antiguos de destino para auditoría
            cursor.execute("SELECT * FROM permisos_perfil WHERE perfil_id = %s", (destino_id,))
            permisos_antiguos = cursor.fetchall()
            
            connection.begin()
            
            # Eliminar permisos antiguos del destino
            cursor.execute("DELETE FROM permisos_perfil WHERE perfil_id = %s", (destino_id,))
            
            # Copiar permisos del origen al destino
            cursor.execute("""
                INSERT INTO permisos_perfil (
                    perfil_id, funcion_id, modulo_id, puede_ejecutar, ver_datos,
                    crear_datos, editar_datos, eliminar_datos, ver_reportes,
                    exportar_datos, importar_datos, imprimir_datos, solo_propios, creado_por
                )
                SELECT %s, funcion_id, modulo_id, puede_ejecutar, ver_datos,
                       crear_datos, editar_datos, eliminar_datos, ver_reportes,
                       exportar_datos, importar_datos, imprimir_datos, solo_propios, %s
                FROM permisos_perfil WHERE perfil_id = %s
            """, (destino_id, usuario_id, origen_id))
            
            # Registrar auditoría
            sql_audit = """
                INSERT INTO auditoria_accesos (
                    usuario_id, tipo_evento, modulo_accedido, funcion_ejecutada,
                    descripcion_evento, direccion_ip, navegador_user_agent,
                    resultado, tabla_afectada, registro_afectado_id,
                    operacion_realizada, valor_anterior
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            cursor.execute(sql_audit, (
                usuario_id,
                'copia_permisos',
                'configuracion',
                'PER_GESTIONAR',
                f"Copiados permisos desde el perfil {perfil_origen['nombre_perfil']} al perfil {perfil_destino['nombre_perfil']}",
                '127.0.0.1',
                'Python Engine API',
                'exitoso',
                'permisos_perfil',
                destino_id,
                'copy',
                json.dumps(permisos_antiguos, default=str)
            ))
            
            connection.commit()
            return {
                'exito': True,
                'mensaje': f"Permisos copiados exitosamente de {perfil_origen['nombre_perfil']} a {perfil_destino['nombre_perfil']}."
            }
    except Exception as e:
        connection.rollback()
        return {'exito': False, 'mensaje': f'Error en copy_profile_permissions: {str(e)}'}
    finally:
        connection.close()

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(json.dumps({'exito': False, 'mensaje': 'Falta el comando principal'}))
        sys.exit(1)
        
    cmd = sys.argv[1]
    
    if cmd == 'list_modules_functions':
        res = list_modules_functions()
    elif cmd == 'get_profile_permissions':
        if len(sys.argv) < 3:
            res = {'exito': False, 'mensaje': 'Falta ID del perfil'}
        else:
            res = get_profile_permissions(sys.argv[2])
    elif cmd == 'save_profile_permissions':
        if len(sys.argv) < 5:
            res = {'exito': False, 'mensaje': 'Faltan parámetros para save_profile_permissions'}
        else:
            res = save_profile_permissions(sys.argv[2], sys.argv[3], sys.argv[4])
    elif cmd == 'create_profile':
        if len(sys.argv) < 4:
            res = {'exito': False, 'mensaje': 'Faltan parámetros para create_profile'}
        else:
            res = create_profile(sys.argv[2], sys.argv[3])
    elif cmd == 'update_profile':
        if len(sys.argv) < 5:
            res = {'exito': False, 'mensaje': 'Faltan parámetros para update_profile'}
        else:
            res = update_profile(sys.argv[2], sys.argv[3], sys.argv[4])
    elif cmd == 'delete_profile':
        if len(sys.argv) < 4:
            res = {'exito': False, 'mensaje': 'Faltan parámetros para delete_profile'}
        else:
            res = delete_profile(sys.argv[2], sys.argv[3])
    elif cmd == 'copy_profile_permissions':
        if len(sys.argv) < 5:
            res = {'exito': False, 'mensaje': 'Faltan parámetros para copy_profile_permissions'}
        else:
            res = copy_profile_permissions(sys.argv[2], sys.argv[3], sys.argv[4])
    else:
        res = {'exito': False, 'mensaje': f'Comando no reconocido: {cmd}'}
        
    print(json.dumps(res, indent=2, default=str))
